xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,271 @@
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+import static io.grpc.xds.XdsLoadBalancerProvider.XDS_POLICY_NAME;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.Attributes;
+import io.grpc.ChannelLogger;
+import io.grpc.ChannelLogger.ChannelLogLevel;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.LoadBalancerRegistry;
+import io.grpc.NameResolver.ConfigOrError;
+import io.grpc.Status;
+import io.grpc.internal.ObjectPool;
+import io.grpc.internal.ServiceConfigUtil.LbConfig;
+import io.grpc.util.GracefulSwitchLoadBalancer;
+import io.grpc.xds.CdsLoadBalancerProvider.CdsConfig;
+import io.grpc.xds.XdsClient.ClusterUpdate;
+import io.grpc.xds.XdsClient.ClusterWatcher;
+import io.grpc.xds.XdsLoadBalancerProvider.XdsConfig;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Load balancer for experimental_cds LB policy.
+ */
+public final class CdsLoadBalancer extends LoadBalancer {
+  private final ChannelLogger channelLogger;
+  private final LoadBalancerRegistry lbRegistry;
+  private final GracefulSwitchLoadBalancer switchingLoadBalancer;
+
+  // The following fields become non-null once handleResolvedAddresses() successfully.
+
+  // Most recent CdsConfig.
+  @Nullable
+  private CdsConfig cdsConfig;
+  // Most recent ClusterWatcher.
+  @Nullable
+  private ClusterWatcher clusterWatcher;
+  @Nullable
+  private ObjectPool<XdsClient> xdsClientRef;
+  @Nullable
+  private XdsClient xdsClient;
+
+  CdsLoadBalancer(Helper helper) {
+    this(helper, LoadBalancerRegistry.getDefaultRegistry());
+  }
+
+  @VisibleForTesting
+  CdsLoadBalancer(Helper helper, LoadBalancerRegistry lbRegistry) {
+    this.channelLogger = helper.getChannelLogger();
+    this.lbRegistry = lbRegistry;
+    this.switchingLoadBalancer = new GracefulSwitchLoadBalancer(helper);
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Received ResolvedAddresses '%s'", resolvedAddresses);
+    Attributes attributes = resolvedAddresses.getAttributes();
+    if (xdsClientRef == null) {
+      xdsClientRef = resolvedAddresses.getAttributes().get(XdsAttributes.XDS_CLIENT_REF);
+      if (xdsClientRef == null) {
+        // TODO(zdapeng): create a new xdsClient from bootstrap if no one exists.
+        throw new UnsupportedOperationException(
+            "XDS_CLIENT_REF attributes not available in resolvedAddresses " + resolvedAddresses);
+      }
+      xdsClient = xdsClientRef.getObject();
+    }
+
+    Map<String, ?> newRawLbConfig = checkNotNull(
+        attributes.get(ATTR_LOAD_BALANCING_CONFIG), "ATTR_LOAD_BALANCING_CONFIG not available");
+    ConfigOrError cfg =
+        CdsLoadBalancerProvider.parseLoadBalancingConfigPolicy(newRawLbConfig);
```

Comment:
I looked at `CdsLoadBalancerProvider.parseLoadBalancingConfigPolicy` : it catches a `RuntimeException` and returns a `ConfigOrError` filled with Error object. And here you are converting it back to `RuntimeException` and throwing it :-) . I guess it's okay if there are other callers of `CdsLoadBalancerProvider.parseLoadBalancingConfigPolicy` that rely on this behavior.

Reply:
Now handled the error by going to TRANSIENT_FAILURE. Once service config error handling is  implemented this will not be a problem anyway - we will never reach it here.